types/src/rsmt2d.rs

```diff
@@ -206,29 +263,33 @@ impl ExtendedDataSquare {
     ) -> Result<Vec<NamespacedData>> {
         let mut data = Vec::new();
 
-        for i in 0u16..self.square_len as u16 {
-            let row_root = dah.row_root(i.into()).unwrap();
+        for row in 0..self.square_len {
+            let Some(row_root) = dah.row_root(row) else {
+                continue;
+            };
+
             if !row_root.contains::<NamespacedSha2Hasher>(*namespace) {
                 continue;
             }
 
             let mut shares = Vec::with_capacity(self.square_len);
-            let mut tree = Nmt::with_hasher(NamespacedSha2Hasher::with_ignore_max_ns(true));
-            for (col, s) in self.row(i.into())?.iter().enumerate() {
-                let ns = if col < self.square_len / 2 {
-                    Namespace::from_raw(&s[..NS_SIZE])?
+
+            for col in 0..self.square_len {
+                let share = self.share(row, col)?;
+
+                let ns = if is_ods_square(row, col, self.square_len) {
```

Comment:
now that I think about it, since the row is ordered by the namespace, we could have something like this after this `let ns ...`:

```rust
if ns > namespace {
    break;
}
```

Reply:
done